src/backend/src/app-mux.jl

```diff
@@ -0,0 +1,209 @@
+@enum FlowDirection Incoming Outgoing
+
+using Statistics
+
+"""
+    get_aggregate_quantiles(data, variable, p, direction::FlowDirection)
+
+Compute the aggregated flows in `direction` for `variable` and compute the quantiles of those flows.
+`p` is a vector of the probabilities at which to calculate the quantiles (see `Statistics.quantile`).
+
+Motivation: chloropleth view.
+"""
+function get_aggregate_quantiles(data, variable, p, direction)
+    vars = mapreduce(vcat, scenarios_with(data, variable)) do scen
+        sum.(get_grouped(data, variable, scen, direction))
+    end
+    return quantile(vars, p)
+end
+
+"""
+    get_aggregate_quantiles(data, variable, p, comparison::Bool, percent::Bool, direction::FlowDirection)
+
+Sample the pairwise comparisons between all scenarios featuring `variable` and compute the sample quantiles.
+`p` is a vector of the probabilities at which to calculate the quantiles (see `Statistics.quantile`).
+
+Motivation: comparison chloropleth view.
+"""
+function get_aggregate_quantiles(data,
+                                 variable,
+                                 p,
+                                 comparison::Bool,
+                                 percent::Bool,
+                                 direction::FlowDirection)
+end
+
+"""
+    get_quantiles(data, variable, p)
+
+Return the quantiles for all values of `variable`.
+`p` is a vector of the probabilities at which to calculate the quantiles (see `Statistics.quantile`).
+
+Motivation: used for scaling flow lines (and, in GCVT, link lines).
+"""
+function get_quantiles(data, variable, p)
+end
+
+"""
+   get_aggregate_flows(data, variable, scenario, direction, zones)
+
+Aggregated (always summed?) flows for all zones. Returns an iterable of pairs (zone_id => flow_amount), maybe a Dict.
+
+Motivation: chloropleth view.
+
+If `zones` is not `:`, only sum flows from/to those zones.
+
+Motivation: chloropleth once you've selected one or more zones.
+"""
+function get_aggregate_flows(data, variable, scenario, direction, zones)
+    col_idx = column_index(data, variable, scenario)
+    if direction == :incoming
+        # Sum only flows originating in `zones`
+        return map(data.grouped_by_destination) do grp
+            sum(filter(:origin => in(zones), grp)[!, col_idx])
+        end
+    else
+        # Sum only flows ending in `zones`
+        return map(data.grouped_by_origin) do grp
+            sum(filter(:destination => in(zones), grp)[!, col_idx])
+        end
+    end
+end
```

Comment:
I think this is noticeably slow. It's also touching the internals of `data`, so some of it shouldn't be here.

Reply:
It's vastly slower than the colon version:

```
julia> @benchmark get_aggregate_flows(data, "sip", "dutch", :incoming, :)
BenchmarkTools.Trial: 
  memory estimate:  206.78 KiB
  allocs estimate:  1432
  --------------
  minimum time:     127.774 μs (0.00% GC)
  median time:      132.663 μs (0.00% GC)
  mean time:        160.176 μs (10.13% GC)
  maximum time:     4.752 ms (92.63% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark get_aggregate_flows(data, "sip", "dutch", :outgoing, :)
BenchmarkTools.Trial: 
  memory estimate:  287.64 KiB
  allocs estimate:  2838
  --------------
  minimum time:     209.166 μs (0.00% GC)
  median time:      216.910 μs (0.00% GC)
  mean time:        266.965 μs (10.86% GC)
  maximum time:     6.377 ms (91.38% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

```
julia> @benchmark get_aggregate_flows(data, "sip", "dutch", :outgoing, $(rand(1:100, 30)))
BenchmarkTools.Trial: 
  memory estimate:  13.31 MiB
  allocs estimate:  90555
  --------------
  minimum time:     21.496 ms (0.00% GC)
  median time:      22.378 ms (0.00% GC)
  mean time:        24.420 ms (4.99% GC)
  maximum time:     37.358 ms (13.03% GC)
  --------------
  samples:          205
  evals/sample:     1

julia> @benchmark get_aggregate_flows(data, "sip", "dutch", :incoming, $(rand(1:100, 30)))
BenchmarkTools.Trial: 
  memory estimate:  6.33 MiB
  allocs estimate:  41099
  --------------
  minimum time:     10.341 ms (0.00% GC)
  median time:      10.787 ms (0.00% GC)
  mean time:        12.284 ms (5.27% GC)
  maximum time:     29.503 ms (19.82% GC)
  --------------
  samples:          408
  evals/sample:     1
```

I'd say it probably is worth trying to optimise.

Reply:
Hmm, it doesn't depend on the number of zones picked: 

```
julia> @benchmark get_aggregate_flows(data, "sip", "dutch", :outgoing, $(rand(1:100, 3)))
BenchmarkTools.Trial: 
  memory estimate:  13.31 MiB
  allocs estimate:  90555
  --------------
  minimum time:     21.557 ms (0.00% GC)
  median time:      22.728 ms (0.00% GC)
  mean time:        25.086 ms (4.90% GC)
  maximum time:     44.953 ms (12.18% GC)
  --------------
  samples:          200
  evals/sample:     1
```

so we're probably just doing something a bit silly somewhere.

Reply:
`@code_warntype` says Julia can't work out what sort of array we're supposed to be returning, AFAICT.

Reply:
I think on balance it's fair that Julia can't work out the right type - `column_index(data, variable, scenario)` can return an index that corresponds to a column of any type. We need to somehow pinky-promise to Julia that we'll only use the function on Float64 columns.

Reply:
```
function get_aggregate_flows(data, variable, scenario, direction, zones)::Array{Float64, 1}
    col_idx = column_index(data, variable, scenario)
    if direction == :incoming
        # Sum only flows originating in `zones`
        return map(data.grouped_by_destination) do grp
            sum(filter(:origin => in(zones), grp)[!, col_idx]::Array{Float64,1})
        end
    else
        # Sum only flows ending in `zones`
        return map(data.grouped_by_origin) do grp
            sum(filter(:destination => in(zones), grp)[!, col_idx]::Array{Float64,1})
        end
    end
end
```

Adding type annotations removes all the red from `@code_warntype` but doesn't improve perf at all. Must have been a dead heron

Reply:
There's loads of allocations, so removing the intermediate filter will probably fix it:

```jl
        return map(data.grouped_by_destination) do grp    
            origins = grp[!, 1]                                                                                              
            col = grp[!, col_idx]                                                                           
            gen = (v for (o, v) in zip(origins, col) if o in zones)                                                                                            
            reduce(+, gen; init = zero(eltype(col)))                                                                                
        end
```

This brings it down to microseconds again for me.

Code is a bit ugly, feel free to improve it.

Reply:
This type assertion is bad, too:

```jl
            sum(filter(:origin => in(zones), grp)[!, col_idx]::Array{Float64,1})
```

It should be:

```jl
            sum(filter(:origin => in(zones), grp)[!, col_idx]::Float64
```

Doesn't improve performance any, tho. DataFrames in general are quite type-unstable by design (to avoid loads of compilation).

Reply:
I don't see how the type assertion could be wrong ... `sum([not an array])` doesn't make much sense. Or do you just have a typo and you're saying it's better to do `sum(...)::Float64`?

I got a bit lost with the type checking because I was in TypeScript land and thinking that `Float64[] === Array{Float64,1}`, which gave me confusing errors.

... anyway, none of this matters. I golfed your solution a bit and pushed it.

Reply:
I was thinking the generator bit could go in ODData.jl as part of one of the getindex methods or something.

I guess I'll have a think later.

Reply:
And yeah, the thing about type annotations was just a misunderstanding and a typo/misreading